### PR TITLE
[HttpClient] doc how to cancel a request + upload files

### LINKED ISSUE #11658: Add docs for: [HttpClient] add $response->cancel()
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#31831
| PR author(s) | @nicolas-grekas

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Fixes #11658. Needs https://github.com/symfony/symfony/pull/31850

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
When merging this into 4.4, don't miss removing the changes done in https://github.com/symfony/symfony-docs/pull/11663

**Comment by nicolas-grekas** [comment]
This can merged now :)

File: components/http_client.rst [hunk 1]
```diff
@@ -303,6 +333,29 @@ response sequentially instead of waiting for the entire response::
...
+
+      $response = $client->request('GET', 'https://..;', [
+        'on_progress' => function (int $dlNow, int $dlSize, array $info): void {
+            // ...
+
+            throw new \MyException();
+        },
+      ]);
+
+  The exception will be wrapped in a ``TransportExceptionInterface`` and will
```
Comment (OskarStark) [inline components/http_client.rst]: This sentence sounds a bit weird, because you cannot wrap sth by an interface but an implementation, right?

File: components/http_client.rst [hunk 2]
```diff
@@ -303,6 +333,29 @@ response sequentially instead of waiting for the entire response::
...
+fetch only the first bytes of the response, etc.), you can either:
+
+* use the ``cancel()`` method of ``ResponseInterface``::
+
+      $response->cancel()
+
+* throw an exception from a progress callback::
+
+      $response = $client->request('GET', 'https://..;', [
+        'on_progress' => function (int $dlNow, int $dlSize, array $info): void {
```
Comment (OskarStark) [inline components/http_client.rst]: Can we rename dlName and dlSize? What does Now mean, size is more clear but.. hmm 🧐😬

File: components/http_client.rst [hunk 3]
```diff
@@ -199,12 +199,39 @@ When uploading data with the ``POST`` method, if you don't define the
 form data and adds the required
 ``'Content-Type: application/x-www-form-urlencoded'`` header for you.
 
-When uploading JSON payloads, use the ``json`` option instead of ``body``. The
-given content will be JSON-encoded automatically and the request will add the
-``Content-Type: application/json`` automatically too::
+When the `body` option is set as a closure, it will be called several times until
```
Comment (OskarStark) [inline components/http_client.rst]: Please 2 backticks around body

File: components/http_client.rst (hunk 1 above)
Comment (nicolas-grekas) [reply components/http_client.rst]: sure, but that's still what you need to catch, the implementation is detail

File: components/http_client.rst (hunk 2 above)
Comment (nicolas-grekas) [reply components/http_client.rst]: that's the name in the code :)
dlNow => downloaded at the very moment this is called
dlSize => the total size of the content (-1 if unkown)
"dl" by opposition with "ul" for upload, since this is called also during upload (then $info contains more details)

Comment (nicolas-grekas) [reply components/http_client.rst]: I'm going to move this in a dedicated section

Comment (nicolas-grekas) [reply components/http_client.rst]: pushed

File: components/http_client.rst [hunk 4]
```diff
@@ -303,6 +348,29 @@ response sequentially instead of waiting for the entire response::
...
+
+      $response = $client->request('GET', 'https://..;', [
+        'on_progress' => function (int $dlNow, int $dlSize, array $info): void {
+            // ...
+
+            throw new \MyException();
+        },
+      ]);
+
+  The exception will be wrapped in an instance of ``TransportExceptionInterface``
```
Comment (OskarStark) [inline components/http_client.rst]: Unfortunately the symfony cloud endpoint for this PR is not bootstrapped yet :(
This belongs to the `* throw an exception from a progress callback::`-Part, but I am not sure if it will be rendered correct with this indention

cc @javiereguiluz

**Review by OskarStark — APPROVED** [review]
Thank you Nicolas 👍

File: components/http_client.rst [hunk 5]
```diff
@@ -232,12 +259,47 @@ making a request. Use the ``max_redirects`` setting to configure this behavior
...
 
+HTTP Proxies
+~~~~~~~~~~~~
+
+By default, this component honors the standard environment variables that your
+Operating System defines to direct the HTTP traffic through your local proxy.
+This means there is usually nothing to configure to have the client work with
+proxies, provided these env vars are properly configured.
+
+You can still set or override these settings using the `proxy`` and ``no_proxy``
```
Comment (javiereguiluz) [inline components/http_client.rst]: There's a missing opening backtick in `proxy`:

```
`proxy`` -> ``proxy``
```

File: components/http_client.rst [hunk 6]
```diff
@@ -232,12 +259,47 @@ making a request. Use the ``max_redirects`` setting to configure this behavior
...
+
+* ``no_proxy`` disables the proxy for a comma-separated list of hosts that do not
+  require it to get reached.
+
+Progress Callback
+~~~~~~~~~~~~~~~~~
+
+By providing a callable to the ``on_progress`` option, one can track
+uploads/downloads as they complete. This callback is guaranteed to be called on
+DNS resolution, on arrival of headers and on completion; additionnaly it is
```
Comment (javiereguiluz) [inline components/http_client.rst]: `additionnaly` -> `additionally`

File: components/http_client.rst (hunk 4 above)
Comment (javiereguiluz) [reply components/http_client.rst]: Yes, while merging we can reword this to remove the `<ul>` list and turn it into paragraphs:

```
To abort a request (e.g. because it didn't complete in due time, or you want to
fetch only the first bytes of the response, etc.), you can either use the ``cancel()`` 
method of ``ResponseInterface``:

   ...

Or throw an exception from a progress callback::

   ...

The exception will be wrapped in an instance of ``TransportExceptionInterface``
and will abort the request.
```

**Review by javiereguiluz — APPROVED** [review]
Very nice! Thanks Nicolas.

File: components/http_client.rst [hunk 7]
```diff
@@ -303,6 +364,29 @@ response sequentially instead of waiting for the entire response::
...
+To abort a request (e.g. because it didn't complete in due time, or you want to
+fetch only the first bytes of the response, etc.), you can either:
+
+* use the ``cancel()`` method of ``ResponseInterface``::
+
+      $response->cancel()
+
+* throw an exception from a progress callback::
+
+      $response = $client->request('GET', 'https://..;', [
```
Comment (andreia) [inline components/http_client.rst]: Just a little detail

```suggestion
      $response = $client->request('GET', 'https://...', [
```

**Review by andreia — CHANGES_REQUESTED** [review]

File: components/http_client.rst (hunk 7 above)
Comment (OskarStark) [reply components/http_client.rst]: Good catch 👍🏻

**Comment by nicolas-grekas** [comment]
Comments addressed, thank you all!

**Review by OskarStark — APPROVED** [review]

**Comment by javiereguiluz** [comment]
Thank you Nicolas.
